README.md
# Feegow Public API

## Importing

#### Node

1. `npm install @feegow/publicapi`
2. `const FeegowPublicAPI = require('@feegow/publicapi');`

#### HTML

1. `<script src="https://unpkg.com/@feegow/publicapi/FeegowPublicAPI.js"></script>`

## Usage

```
const feegowApi = new FeegowPublicAPI('<your_x_access_token>');

// Calls an endpoint without specifying version or passing parameters
feegowApi.getPatients();

// Calls an endpoint specifying version
feegowApi.getPatients('v1');

// Calls an endpoint specifying version and passing parameters
feegowApi.getPatients('v1', {
    "view_mode": "full", // not working
    "limit": 1, // not working
    "offset": 0,
    "telefone": "(21) 2018-0123",
    "cpf": "177.820.767-73"
});
```

## Publicando uma nova versão

- Documente as alterações da versão no `CHANGELOG.md`
- Incremente a versão no `package.json`
- Execute `npm run pub-push`
- Adicione a versão no campo `Release title` da página que abrir

> Ao executar `npm run pub-push` são executados os comandos abaixo:
> ```
> npm publish
> git add .
> git commit -m \"Update %npm_package_version%\"
> git tag %npm_package_version%
> git push origin
> git push origin %npm_package_version%
> start https://github.com/feegow/feegow-public-api-node-package/releases/new?tag=%npm_package_version%
> ```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.0.15](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.14...v1.0.15) (2021-07-30)

- Fix setting wrong path on endpoint

## [1.0.14](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.13...v1.0.14) (2021-07-30)

- Organização das branchs do repositório

## [1.0.13](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.12...v1.0.13) (2021-07-30)

- Fix `FeegowPublicAPI.js:52 GET https://api.feegow.com/v1/api/patient/list?offset=null&limit=null&telefone=null&cpf=null&view_mode=null 422`

## [1.0.12](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.11...v1.0.12) (2021-07-30)

- Fix `ReferenceError: myHeaders is not defined`

## [1.0.11](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.10...v1.0.11) (2021-07-30)

- Fix CORS

## [1.0.10](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.9...v1.0.10) (2021-07-30)

- Fix CORS

## [1.0.9](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.8...v1.0.9) (2021-07-30)

- Add NPM script push-publish

## [1.0.8](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.7...v1.0.8) (2021-07-30)

- Fix CORS

## [1.0.7](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.6...v1.0.7) (2021-07-30)

- Fix CORS

## [1.0.6](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.5...v1.0.6) (2021-07-30)

- Replacement of axios for node-fetch
- Add documentation with example

## [1.0.5](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.4...v1.0.5) (2021-07-30)

- Error: Cannot find module 'C:\Dev\feegow-public-api-node-package\testeee\node_modules\@feegow\publicapi\index.js'. Please verify that the package.json has a valid "main" entry

## [1.0.4](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.3...v1.0.4) (2021-07-30)

- Uncaught SyntaxError: Cannot use import statement outside a module

## [1.0.3](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.2...v1.0.3) (2021-07-30)

- FeegowPublicAPI.js:1 Uncaught ReferenceError: require is not defined

## [1.0.2](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.1...v1.0.2) (2021-07-30)

- Uncaught SyntaxError: Identifier 'FeegowPublicAPI' has already been declared

## [1.0.1](https://github.com/feegow/feegow-public-api-node-package/compare/v1.0.0...v1.0.1) (2021-07-30)

- Dotenv support in test file.
- Import by `<script src="https://unpkg.com/@feegow/publicapi@1.0.1/FeegowPublicAPI.js"></script>`

## [1.0.0](https://github.com/feegow/feegow-public-api-node-package/tree/1.0.0) (2021-07-22)

- Published package.

test/node/index.js
const FeegowPublicAPI = require('@feegow/publicapi');

const a = new FeegowPublicAPI('aefaefaf');

src/FeegowPublicAPI.test.js
require('dotenv').config();

const FeegowPublicAPI = require('./FeegowPublicAPI');

var feegowApi;

test('Authenticate with valid token', () => {

    feegowApi = new FeegowPublicAPI(process.env.TOKEN);

    const tokenInfo = feegowApi.tokenInfo;

    //console.log(tokenInfo);

    expect(tokenInfo).toHaveProperty('iss');
    expect(tokenInfo).toHaveProperty('aud');
    expect(tokenInfo).toHaveProperty('iat');
    expect(tokenInfo).toHaveProperty('licenseID');
});

test('Authenticate with invalid token', () => {
    expect(() => {
        feegowApi = new FeegowPublicAPI('aefaefaefaef');
    }).toThrow('Token inválido');
});

test('Get all patients', async () => {
    await expect(feegowApi.getPatients()).resolves.toHaveProperty('success', true)
}, 10000);


test/FeegowPublicAPI.test.js
require('dotenv').config();

const FeegowPublicAPI = require('../FeegowPublicAPI');

var feegowApi;

test('Authenticate with valid token', () => {

    feegowApi = new FeegowPublicAPI(process.env.TOKEN);

    const tokenInfo = feegowApi.tokenInfo;

    //console.log(tokenInfo);

    expect(tokenInfo).toHaveProperty('iss');
    expect(tokenInfo).toHaveProperty('aud');
    expect(tokenInfo).toHaveProperty('iat');
    expect(tokenInfo).toHaveProperty('licenseID');
});

test('Authenticate with invalid token', () => {
    expect(() => {
        feegowApi = new FeegowPublicAPI('aefaefaefaef');
    }).toThrow('Token inválido');
});

test('Get all patients', async () => {
    await expect(feegowApi.getPatients()).resolves.toHaveProperty('success', true)
}, 10000);


FeegowPublicAPI.js

if (typeof exports != "undefined") {  
    var { atob } = require('atob');
    var fetch = require('node-fetch');
}

class FeegowPublicAPI {

    constructor(token) {

        this.getPatients = this.getPatients.bind(this);
        this.checkAndFormatVersion = this.checkAndFormatVersion.bind(this);

        const tokenInfo = this.decryptJwt(token);

        if (!tokenInfo) {
            throw 'Token inválido';
        }

        this.token = token;
        this.tokenInfo = tokenInfo;
    }

    /**
     * @example 
     * const data = {
     *    "view_mode": "full", // not working
     *    "limit": 1, // not working
     *    "offset": 0,
     *    "telefone": "(21) 2018-0123",
     *    "cpf": "177.820.767-73"
     * };
     * */
    getPatients(version = 'v1', data = {}) {

        version = this.checkAndFormatVersion(version);
        
        return new Promise((resolve, reject) => {

            const filter = new URLSearchParams(data).toString();

            const headers = new Headers();
            headers.append("X-Access-Token", this.token);
            headers.append("Content-Type", "application/json");
    
            fetch(` https://api.feegow.com/v1/api/${version}patient/list?${filter}`, { 
                method: 'GET',
                headers: headers
            })
                .then(res => res.json())
                .then(json => resolve(json))
                .catch((error) => reject(error['response'] ? error['response']['data'] : error));
        });
    }

    decryptJwt(token) {
        try {
            const base64Url = token.split('.')[1];
            const base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
            const jsonPayload = decodeURIComponent(atob(base64).split('').map(function (c) {
                return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
            }).join(''));
            return JSON.parse(jsonPayload);
        } catch (err) {
            return false;
        }
    }

    checkAndFormatVersion(version) {
        switch(version) {
            case 'v1':
            case 'V1':
                return '';
            case 'v2':
            case 'V2':
                return 'v2/';
            case 'v3':
            case 'V3':
                return 'v3/';
            default:
                throw 'Unknown API version.';
        }
    }
}

if (typeof exports != "undefined") {    
    module.exports = FeegowPublicAPI;
}